### PR TITLE
Feat: Add app upgrade

### PR DESCRIPTION
**Pull request by niristius**
This feature is not yet done.
There are still some problems with choosing specific target versions for an upgrade.
Also prompts, user interaction aswell as general code efficiency and hygiene needs some more work.
Any feedback is already welcome though!

### CODE REVIEW

File: src/commands/app/upgrade.tsx
Comment (martin-helmich) [inline src/commands/app/upgrade.tsx:35]: Listing potential target versions and actually performing an update are quite different things; maybe extract this functionality into a different command? Like `mw app list-upgrade-candidates`?

EDIT: That command would also make sense to be implemented as a `ListBaseCommand` subclass, because I could imagine a user wanting JSON/YAML output for that command (which would not be easily possible with the current implementation).

File: src/commands/app/upgrade.tsx
Comment (martin-helmich) [inline src/commands/app/upgrade.tsx:60]: `assertStatus` to assert the expected return status code; this should also make the subsequent type casts unnecessary, because the compiler should be able to figure out the data types on its own.

File: src/commands/app/upgrade.tsx
Comment (martin-helmich) [inline src/commands/app/upgrade.tsx:138]:
```suggestion
          Confirm upgrading {currentApp.name}{" "}
          {currentAppVersion.externalVersion} (description here) to version{" "}
          {targetAppVersion.externalVersion}
```

https://www.grammarly.com/blog/capitalization-rules/ 😉

File: src/commands/app/upgrade.tsx
Comment (martin-helmich) [inline src/commands/app/upgrade.tsx:151]:
```suggestion
          {currentAppVersion.externalVersion} (description here) to version
```

File: src/commands/app/upgrade.tsx
Comment (martin-helmich) [inline src/commands/app/upgrade.tsx:164]: Maybe also add a `--wait` flag (like for the `install [appname]` commands), in case someone wants to use this command synchronously?

File: src/lib/app/versions.tsx
Comment (martin-helmich) [inline src/lib/app/versions.tsx:87]: If we're using semver functionalities anyway: Should we just offer the option for the user to specify a semver selector? Like...

```
$ mw app upgrade <app-id> --target-version="^12.4.0"
$ mw app upgrade <app-id> --target-version="*"
```

Consequently, the `install` commands should then support this, as well.

Could also be implemented as a follow-up PR.

File: src/commands/app/upgrade.tsx
Comment (martin-helmich) [reply src/commands/app/upgrade.tsx:138]: Also, what's up with `description here`?

**Review by martin-helmich — CHANGES_REQUESTED** [review]

File: src/commands/app/upgrade.tsx
Comment (niristius) [reply src/commands/app/upgrade.tsx:35]: Valid point - This will be implemented.
I would argue though, that since most of the same data is required for the upgrade process anyway, so giving the user the option to check the candidates while already preparing the upgrade command might make for a smoother workflow - Albeit this means that either the flag or the extra command would be redundant. I don't see the harm though except some degree of bloating the cli?

Comment (martin-helmich) [reply src/commands/app/upgrade.tsx:35]: Or, alternatively, while thinking about it some more: If no `--target-version` was specified, give the user the option to select the desired target version interactively (if used within a TTY context) from the candidate list; this could be done using the `ProcessRenderer`s `addSelect` method:

https://github.com/mittwald/cli/blob/d8076f7877613b86aaebabeb8bf55654003245fd/src/rendering/process/process.tsx#L116-L119

File: src/commands/app/upgrade.tsx
Comment (niristius) [reply src/commands/app/upgrade.tsx]: mentioned typos have been corrected, and then some.

File: src/lib/app/versions.tsx
Comment (niristius) [reply src/lib/app/versions.tsx:105]: I like this.
I'll add it as a todo for this PR, but we'll see if we don't just merge somewhere in between already and make a new PR with less impactful stuff

File: src/commands/app/upgrade.tsx
Comment (niristius) [reply src/commands/app/upgrade.tsx]: it's on the todo list, just wanted to get basic functionalities out of the way first and then look into how the waiting at that point actually works before I just copy&paste functionalities.

File: src/commands/app/upgrade.tsx
Comment (niristius) [reply src/commands/app/upgrade.tsx:125]: fixed in 25a4535f81bd2aa692f974e7819dc733581dab6c

File: src/commands/app/upgrade.tsx
Comment (niristius) [reply src/commands/app/upgrade.tsx]: I'll have to keep an eye out for correct assertion, and will do that.
For this explicit instance the whole api call was put into a wrapper function with correct assertion in 2fe9d63108be465704faa55aae270c26dfb4f3ce

**Comment by niristius** [comment]
Still Todo

- [ ] fix up the "list-appupgrade-candidates" command to actually make it work
- [ ] add app installation description to user dialogue where it make sense
- [ ] make appinstallation selection interactive if none is given
- [ ] make upgrade target version selection interactive if none given
- [ ] add "--wait" flag 
- [ ] target version semver functionalities

**Comment by niristius** [comment]
This PR is ready to be merged after having gotten a second opinion on user dialogue and general functionality.
I merged every commit that was pushed into master since this branch was created and generated the readme but the readme check still fails, what am I missing?

About the two open todos:

Making the Project or App as an interactive selection would need a bit of work on the "withProjectId"/"withAppInstallationId" functions as to not bloat or overcomplicate things. I dont see it as a necessary part of this PR's feature.

I will tackle target version semver functionalities in a follow up PR at a later date.

File: README.md
Comment (martin-helmich) [inline README.md:1999]: Looks like you somehow DOWNgraded some of the dependencies. Since the `yarn.lock` is unmodified → did you perhaps forget to install the latest dependencies before generating the README?

File: src/commands/app/list-upgrade-candidates.ts
Comment (martin-helmich) [inline src/commands/app/list-upgrade-candidates.ts:24]: Wording consistency: "upgrade" vs. "update"

```suggestion
  static description = "List upgrade candidates for an app installation.";
```

File: src/commands/app/list-upgrade-candidates.ts
Comment (martin-helmich) [inline src/commands/app/list-upgrade-candidates.ts:49]: Why is this typecast necessary?

File: src/commands/app/upgrade.tsx
Comment (martin-helmich) [inline src/commands/app/upgrade.tsx:208]: Do we actually need this special "latest" option? The actual latest version should be the first in the candidate list anyway, shouldn't it?

File: src/lib/app/uuid.ts
Comment (martin-helmich) [inline src/lib/app/uuid.ts:65]: tbh., I'd just rely on the API's response code here, instead of manually validating the input query. If it's not a valid installation ID, the API will tell you soon enough. 🤷

File: src/lib/app/versions.tsx
Comment (martin-helmich) [inline src/lib/app/versions.tsx:36]: Capitalization was correct before.

```suggestion
      )} version ${version} does not seem to exist for the mStudio.`,
```

File: src/lib/app/versions.tsx
Comment (martin-helmich) [inline src/lib/app/versions.tsx:42]: Capitalization was correct before.

```suggestion
      installing version: <Value>{appVersion.externalVersion}</Value>
```

File: src/lib/app/versions.tsx
Comment (martin-helmich) [inline src/lib/app/versions.tsx:111]: Can be simplified; just store the entire version object as `latestVersion` and save yourself the lookup at the end:

```suggestion
  let latestVersion: AppVersion|undefined;
  for (const version of versions.data) {
    if (gt(version.internalVersion, latestVersion?.internalVersion ?? "0.0.0")) {
      latestVersion = version;
    }
  }
  return latestVersion;
```

File: src/lib/app/wait.tsx
Comment (martin-helmich) [inline src/lib/app/wait.tsx:33]: Apart from the label text, this function is identical to `waitUntilAppIsInstalled`. This is a lot of duplication and should be simplified.

File: src/commands/app/upgrade.tsx
Comment (martin-helmich) [inline src/commands/app/upgrade.tsx:36]:
```suggestion
  static description = "Upgrade app installation to target version";
```

File: src/commands/app/upgrade.tsx
Comment (martin-helmich) [inline src/commands/app/upgrade.tsx:43]: This this flag is neither required nor has a default value, the documentation should explain what happens when the flag is omitted.

```suggestion
    "target-version": Flags.string({
      description: "target version to upgrade app to; if omitted, target version will be prompted interactively",
    }),
```

File: src/commands/app/upgrade.tsx
Comment (martin-helmich) [inline src/commands/app/upgrade.tsx:72]: Why is the `as string` typecast necessary?

File: src/commands/app/upgrade.tsx
Comment (martin-helmich) [inline src/commands/app/upgrade.tsx:77]: Why is the `as { current: string }` typecast necessary?

**Review by martin-helmich — CHANGES_REQUESTED** [review]

File: README.md
Comment (niristius) [reply README.md]: I was not aware of this, fixed with 5b1b3a4230e7dd732015d94c52e86267b50156bb

File: src/commands/app/list-upgrade-candidates.ts
Comment (niristius) [reply src/commands/app/list-upgrade-candidates.ts]: It isn't that is a inadequate quickfix I forgot to properly handle because I didn't know how at that time.
Follow up question to that one though: The problem, wich I didnt understand at that time, was just that `baseAppVersionId` expects a string, and `currentAppInstallation.appVersion.current` might be `undefined`. So the solutaion is a basic check if it is undefined.

Now to my question: Usually I would use `process.error()` and let the process handle the error.
In the `ListBaseCommand` we dont have a process though. Do I just `throw.error()`? (90d3c58abce4f3cc34817428761ae64beeac3e6a)

Comment (martin-helmich) [reply src/commands/app/list-upgrade-candidates.ts]: > So the solution is a basic check if it is undefined.

Please note that the typecast does not _actually_ check if `.current` is undefined; you just tell the compiler that it's not (and it will fail at runtime, if it turns out to be undefined, anyway).

You could either assert that `.current` is indeed a string (`typeof currentAppInstallation.appVersion.current === "string"` should be sufficient for the compiler) and throw an error if not (the error will be handled properly by the framework itself);

Alternatively, you might also check against `.desired`, which IRRC is not-nullable -- which might just work better anyway, in case a user decides to invoke `list-upgrade-candidates` on an app installation for which an upgrade is currently in progress.

File: src/commands/app/upgrade.tsx
Comment (niristius) [reply src/commands/app/upgrade.tsx:208]: You aren't wrong - I just thought that giving the latest option would be more in unison with the option to declare the `--target-version` as `latest`, this might be personal preference though.

Also, the list returned by `apiClient.app.listUpdateCandidatesForAppversion()` is not ordered by semver.

- Ordering it during runtime for the interactive selection is not a big task but I'd argue its the wrong layer to do it on.
- Ordering it for the `list-upgrade-candidates`command might be though since this is a direct digest of the api-client-response. 
- Ordering it in the api-client before returning it might be a weird way to handle it since the api-client should contain as little logical digestion as possible.

I would argue this should already be returned by the api as a semver sorted list.
And as soon as the list os actually ordered, the `latest` entry might as well be purged.

File: src/lib/app/uuid.ts
Comment (niristius) [reply src/lib/app/uuid.ts:65]: My train of thought was that this function has no idea where the input comes from, it just gets passed a `string`.
If the given parameters came from an api-response or not should not be relevant for this function explicitly at the time of that check - or is that too petty?

File: src/lib/app/wait.tsx
Comment (niristius) [reply src/lib/app/wait.tsx]: Oh yeah, you're completely right.
I propose using one "waitUntilAppStateHasNormalized"-function with one additional parameter wich is the label.
ab1ab4b251d7b27c441b34889709bbc39299a3f6

File: src/commands/app/upgrade.tsx
Comment (niristius) [reply src/commands/app/upgrade.tsx]: As above, this is a workaround since I didnt understand that its just a missing check if it even is undefined.
Fixed with a check and exit with error code 1 if its undefined.

File: src/commands/app/upgrade.tsx
Comment (niristius) [reply src/commands/app/upgrade.tsx]: It's not. my personal skill issue, I assume. its gone.

File: src/commands/app/upgrade.tsx
Comment (niristius) [reply src/commands/app/upgrade.tsx]: sort implemented

File: src/lib/app/uuid.ts
Comment (niristius) [reply src/lib/app/uuid.ts:65]: Oh, I've only just figured out how you meant that. Feed the string into the getAppinstallation request and see if the client/api can work with the string. Consider it done!

File: src/commands/app/upgrade.tsx
Comment (martin-helmich) [inline src/commands/app/upgrade.tsx:237]: If `latest` is not a selectable option anymore, this condition shouldn't be necessary anymore, either.

```suggestion
  return await getAvailableTargetAppVersionFromExternalVersion(
    apiClient,
    currentApp.id,
    currentAppVersion.id,
    targetAppVersionString,
  );
```

Comment (niristius) [reply src/commands/app/upgrade.tsx:237]: latest was removed from the interactive selection only.
You can still trigger an upgrade to `latest` using the  `--target-version` flag (as of now).

Comment (niristius) [reply src/commands/app/upgrade.tsx:237]: Should I remove it here aswell? I thought it would be a helpful option especially for custom update scripts.

Comment (niristius) [reply src/commands/app/upgrade.tsx:237]: @martin-helmich delete or keep the option to use "latest" for the "target-version"-flag - what's your opinion?
As soon as that is decided eventual changes take about a minute - after that, this feature can be merged.

Comment (martin-helmich) [reply src/commands/app/upgrade.tsx:237]: The option to select the latest version via flag is completely valid. But... this function appears to *only* cover interactive selection (the first statement is an unconditional `await process.addSelect`). If `latest` is not an option for interactive selection, I see no way how `targetAppVersionString` could have `"latest"` as value.

**Review by martin-helmich — APPROVED** [review]
LGTM now. :+1:   (only watch out for the merge conflict)

File: src/commands/app/upgrade.tsx
Comment (niristius) [reply src/commands/app/upgrade.tsx]: I was not paying sufficient attention to the line numbers. 
You are completely right and I need to be more careful.

Fixed it immediately.

**Review by martin-helmich — APPROVED** [review]

**Pull request merged**
